apps/aecore/src/aec_sync.erl

```diff
@@ -685,17 +685,16 @@ post_blocks(From, To, [{Height, _Hash, {PeerId, Block}} | Blocks]) ->
 %% smaller chuncks, a few micro blocks at the time.
 %% Each micro block has a fixed maximum gas, by limiting the number of micro
 %% blocks we limit the total amount of work the conductor has to perform in
-%% each synchronous call. 
+%% each synchronous call.
+%% Map contains key dir, saying in which direction we sync
 add_generation(#{micro_blocks := MBlocks} = Generation) ->
-    add_generation(maps:without([micro_blocks], Generation), MBlocks).
-
--define(MAX_NR_MICROBLOCKS, 4).
-add_generation(Gen, MBlocks) when length(MBlocks) < ?MAX_NR_MICROBLOCKS ->
-    aec_conductor:add_synced_generation_batched(Gen#{micro_blocks => MBlocks}, true);
-add_generation(Gen, MBlocks) ->
-    {Add, AddLater} = lists:split(?MAX_NR_MICROBLOCKS, MBlocks),
-    case aec_conductor:add_synced_generation_batched(Gen#{micro_blocks => Add}, false) of
-        ok -> add_generation(Gen, AddLater);
+    add_micro_blocks_in_generation(Generation#{micro_blocks => []}, MBlocks).
+
+add_micro_blocks_in_generation(Gen, []) ->
+    aec_conductor:add_synced_generation_batched(Gen, true);
```

Comment:
So this is adding the key block last for both backward and forward. This is a change of semantics in how forward generation is added. For forward seems either wrong or redundant as it adds microblocks first then it adds preceding keyblock, but if keyblock insertion was useful then then insertion of all microblocks would have failed as not connected to genesis / orphan. Could you please clarify?

Reply:
The problem we solve here is that generations can have many micro blocks and we need to put each micro block into the database *separately*. That is, it takes time for each micro block and if we do that in 1 call to conductor, then we get timeouts.

We can get generations via gossip and via sync. One is forward, the other backward, meaning that one starts with a key block and then all micro blocks before it, the other all micro blocks after the key block. It is ok to add the same key block twice and not expensive to do so. The code is considered clearer if we don't distinguish forward and backward at this level.... hence the way of writing it.

Reply:
So, for forward it is redundant, but the code is much nicer and it does not cost.

Reply:
Ok.